openapi-bal-service/src/main/java/io/ballerina/openapi/converter/service/OpenAPIComponentMapper.java

```diff
@@ -233,32 +233,49 @@ private Schema handleTypeReference(Map<String, Schema> schema, TypeReferenceType
      *     };
      * </pre>
      */
-    private Schema handleUnionType(Map<String, Schema> schema, Map.Entry<String, RecordFieldSymbol> field,
-                                   Schema property) {
-
-        List<TypeSymbol> unionTypes = ((UnionTypeSymbol) field.getValue().typeDescriptor()).userSpecifiedMemberTypes();
-        String componentName = null;
+    private Schema handleUnionType(UnionTypeSymbol unionType, Schema property) {
+        List<TypeSymbol> unionTypes = unionType.userSpecifiedMemberTypes();
+        //Set array size to 4 by assuming union type can have max 4 types.
+        List<Schema> properties = new ArrayList<>(4);
         boolean nullable = false;
         for (TypeSymbol union: unionTypes) {
             if (union.typeKind() == TypeDescKind.NIL) {
                 nullable = true;
             } else if (union.typeKind() == TypeDescKind.TYPE_REFERENCE) {
                 property = ConverterCommonUtils.getOpenApiSchema(union.typeKind().getName().trim());
                 TypeReferenceTypeSymbol typeReferenceTypeSymbol = (TypeReferenceTypeSymbol) union;
-                property = handleTypeReference(schema, typeReferenceTypeSymbol, property);
-                componentName = union.getName().orElseThrow(null);
+                property = handleTypeReference(this.components.getSchemas(), typeReferenceTypeSymbol, property);
+                properties.add(property);
+                // commented due to known issue in ballerina lang union type handling
+//            } else if (union.typeKind() == TypeDescKind.UNION) {
+//                property = handleUnionType((UnionTypeSymbol) union, property, properties);
+//                properties.add(property);
```

Comment:
Have you created an issue for this? If so, better to add a Todo item to enable this once the fix is merged

Reply:
Yeah There is created issue, will add TODO as well